src/hooks/protections/receive_net_message.cpp

```diff
@@ -98,227 +324,439 @@ namespace big
 			}
 		}
 
-		if (!get_msg_type(msgType, buffer))
-			return g_hooking->get_original<hooks::receive_net_message>()(netConnectionManager, a2, frame);
+		int sec_id = 0;
 
+		if (frame)
+			sec_id = frame->m_security_id;
+		else
+		{
+			if (auto cxn = g_pointers->m_gta.m_get_connection_peer(net_cxn_mgr, error->m_peer_id))
+				sec_id = cxn->m_security_id;
+		}
+
+		auto peer = g_pointers->m_gta.m_get_peer_by_security_id(sec_id); // shouldn't be null in most cases, contains unspoofable data
 
-		if (msgType == rage::eNetMessage::MsgRoamingJoinBubbleAck)
+		if (error && msgType != rage::eNetMessage::MsgJoinResponse)
+		{
+			if (peer)
+				LOGF(stream::net_messages, WARNING, "Received an error packet that isn't MsgJoinResponse from {}", peer->m_info.name);
 			return true;
+		}
 
-		if (player)
+		if (g.debug.logs.packet_logs)
 		{
-			switch (msgType)
+			log_net_message(msgType, buffer, event, peer);
+		}
+
+		switch (msgType)
+		{
+		case rage::eNetMessage::MsgScriptJoin:
+		{
+			CGameScriptId script;
+			script_id_deserialize(script, buffer);
+
+			if (player && handle_block_script(player, script, event->m_msg_id))
+			{
+				LOGF(stream::net_messages, WARNING, "Denying script request from {} (hash={:X}, instance={})", player->get_name(), script.m_hash, script.m_instance_id);
+				return true;
+			}
+
+			break;
+		}
+		case rage::eNetMessage::MsgScriptHostRequest:
+		{
+			CGameScriptId script;
+			script_id_deserialize(script, buffer);
+
+			if (script.m_hash == "freemode"_J && g.session.force_script_host)
+				return true;
+
+			break;
+		}
+		case rage::eNetMessage::MsgKickPlayer:
+		{
+			KickReason reason = buffer.Read<KickReason>(3);
+
+			if (!is_host_of_session(gta_util::get_network()->m_game_session_ptr, event->m_peer_id))
```

Comment:
should be using `session` pointer defined above instead of `gta_util::get_network()->m_game_session_ptr` ?